### PR TITLE
Fix ICE related to receive function having parameters.

### LINKED ISSUE #10882: [OverrideChecker] ICE due to override having different signature
## Description

```
contract C {
  receive(bytes2) {}
}
contract D is C {
  receive() {}
}
```

throws

https://github.com/ethereum/solidity/blob/358324ed3e261dd08ef39e947fc3d4b393427573/libsolidity/analysis/OverrideChecker.cpp#L583

## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ solc test.sol
```

### PR DESCRIPTION
**Pull request by hrkrshnn**
Closes https://github.com/ethereum/solidity/issues/10882

### CODE REVIEW

File: libsolidity/analysis/ContractLevelChecker.cpp [hunk 1]
```diff
@@ -86,6 +86,7 @@ bool ContractLevelChecker::check(ContractDefinition const& _contract)
 
 	checkDuplicateFunctions(_contract);
 	checkDuplicateEvents(_contract);
+	checkReceiveParameters(_contract);
 	m_overrideChecker.check(_contract);
```
Comment (hrkrshnn) [inline libsolidity/analysis/ContractLevelChecker.cpp:90]: The problem is that Override Checker won't be able to handle the example from the issue.

The ICE happens at

https://github.com/ethereum/solidity/blob/4cbf9ff74c9aabe6047212cb60a28728e66b45aa/libsolidity/analysis/OverrideChecker.cpp#L583

At the following point, the two override functions are set to be compared.

https://github.com/ethereum/solidity/blob/4cbf9ff74c9aabe6047212cb60a28728e66b45aa/libsolidity/analysis/OverrideChecker.cpp#L840

**Comment by chriseth** [comment]
The checks in https://github.com/ethereum/solidity/pull/11475/files#diff-1286fd2f351e6a40b993189fde18c5db99ddfe4a35f4ada894945ba62819a324R1884 are all not related to types. Is it possible to move all of them to ContractLevelChecker?

**Comment by hrkrshnn** [comment]
> The checks in https://github.com/ethereum/solidity/pull/11475/files#diff-1286fd2f351e6a40b993189fde18c5db99ddfe4a35f4ada894945ba62819a324R1884 are all not related to types. Is it possible to move all of them to ContractLevelChecker?

@chriseth The link is broken for me. Do you just want to move `TypeChecker::typeCheckReceiveFunction` to ContractLevelChecker? 

It could be possible to move some of `typeCheckFallBackFunction` as well.

File: Changelog.md [hunk 2]
```diff
@@ -23,6 +23,7 @@ Bugfixes:
  * Source Locations: Properly set source location of scoped blocks.
  * Standard JSON: Properly allow the ``inliner`` setting under ``settings.optimizer.details``.
  * Type Checker: Fix internal compiler error related to having mapping types in constructor parameter for abstract contracts.
+ * Type Checker: Fix internal compiler error related to receive ether function having parameters during inheritance.
```
Comment (cameel) [inline Changelog.md]: How about this?
```suggestion
 * Type Checker: Fix internal compiler error when overriding receive ether function with one having different parameters.
```

File: test/libsolidity/syntaxTests/inheritance/fallback_receive/receive_paramter.sol [hunk 3]
```diff
@@ -0,0 +1,10 @@
+contract C {
```
Comment (cameel) [inline test/libsolidity/syntaxTests/inheritance/fallback_receive/receive_paramter.sol]: Typo in test name: `receive_paramter.sol` -> `receive_parameter.sol`

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -1898,10 +1898,6 @@ void TypeChecker::typeCheckReceiveFunction(FunctionDefinition const& _function)
 		);
 	if (_function.visibility() != Visibility::External)
 		m_errorReporter.typeError(4095_error, _function.location(), "Receive ether function must be defined as \"external\".");
-	if (!_function.returnParameters().empty())
-		m_errorReporter.typeError(6899_error, _function.returnParameterList()->location(), "Receive ether function cannot return values.");
-	if (!_function.parameters().empty())
-		m_errorReporter.typeError(6857_error, _function.parameterList().location(), "Receive ether function cannot take parameters.");
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp:1904]: Can you add assertions in place of the removed checks? The only reason we do not have to do them is that they should get caught earlier, right?

File: test/libsolidity/syntaxTests/inheritance/fallback_receive/receive_paramter.sol [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+    receive(bytes2) {}
+}
+contract D is C {
+    receive() {}
+}
```
Comment (cameel) [inline test/libsolidity/syntaxTests/inheritance/fallback_receive/receive_paramter.sol]: Could you also add a similar test case but with return parameters?
```solidity
contract C {
    receive() external payable virtual returns(uint) {}
}
contract D is C {
    receive() external payable override {}
}
```

and one where this happens to an unimplemented function?
```solidity
interface I {
    receive(bytes2) external payable;
}

interface J is I {
    receive() external payable override;
}

contract C is J {
    receive() external payable override {}
}
```

**Comment by chriseth** [comment]
@hrkrshnn Yes, I have the impression that all checks in `TypeChecker::typeCheckReceiveFunction` can be moved because they are unrelated to types.

File: Changelog.md (hunk 2 above)
Comment (hrkrshnn) [reply Changelog.md]: Fixed.

File: libsolidity/analysis/TypeChecker.cpp (hunk 4 above)
Comment (hrkrshnn) [reply libsolidity/analysis/TypeChecker.cpp:1904]: I moved the entire function to ContractLevelChecker now!

File: test/libsolidity/syntaxTests/inheritance/fallback_receive/receive_paramter.sol (hunk 5 above)
Comment (hrkrshnn) [reply test/libsolidity/syntaxTests/inheritance/fallback_receive/receive_paramter.sol]: Added the two tests

**Review by chriseth — APPROVED** [review]

**Pull request merged**